### PR TITLE
Added constructor tests for Touch and TouchEvent

### PR DESCRIPTION
**Pull request by chong-z**
@RByers Added constructor tests for Touch and TouchEvent based on [Touch Events v2 - Touch Interface](https://w3c.github.io/touch-events/#touch-interface) and [Touch Events v2 - TouchEvent Interface](https://w3c.github.io/touch-events/#touchevent-interface)

Browser support:
Chrome Canary: Yes
Firefox, Safari, Edge: Not yet

Issue tracking: https://github.com/w3c/touch-events/issues/27

[Review on Reviewable](https://reviewable.io/reviews/w3c/web-platform-tests/2299)

### CODE REVIEW

**Comment by RByers** [comment]
Thanks for doing this!  Just a couple small concerns.

I've actually never contributed directly to web-platform-test before (I worked on these tests when they were hosted directly in the webevents repository), so I'm not completely sure what the process is.  Looks like there's some information [here](https://github.com/w3c/web-platform-tests/blob/master/README.md).

Please send a note to public-touchevents mentioning this pull request (pull requests to the touchevents repo automatically generate a message to the list, but this is different) so that anyone else in the group can take a look.

I think we should [land the feature in blink](https://code.google.com/p/chromium/issues/detail?id=508675) and verify these tests pass in Chrome Canary before landing the new tests.

**Comment by RByers** [comment]
> Please send a note to public-touchevents mentioning this pull request (pull requests to the touchevents repo automatically generate a message to the list, but this is different) so that anyone else in the group can take a look.

Actually my comment on the issue should be sufficient (looks like all issue comments get copied to the list).

**Comment by AFBarstow** [comment]
This issue is also logged as <https://github.com/w3c/touch-events/issues/27>

**Comment by jgraham** [comment]
To get these merged someone needs to review them. @RByers - are you doing that? Then, ideally, the commits would be squashed, then we can merge.

**Comment by RByers** [comment]
@jgraham Yes I reviewed them back in Nov, but I don't have much experience with WPT so I assumed we'd need review from someone else too.  @choniong, you had talked to someone on #testing about getting a review right?

**Comment by RByers** [comment]
Review status: 0 of 3 files reviewed at latest revision, 1 unresolved discussion.

---

*[touch-events/touch-touchevent-constructor.html, line 15 \[r3\]](https://reviewable.io:443/reviews/w3c/web-platform-tests/2299#-K6E46sdM8kDaMFmRH7D:-K6E46sdM8kDaMFmRH7E:-138903121) ([raw file](https://github.com/w3c/web-platform-tests/blob/cfc9ade7ff7043ea24255bdc2908c3f7777542a4/touch-events/touch-touchevent-constructor.html#L15)):*
Please update this comment.  None of the previous 3 paragraphs applies to this test.  Maybe instead just say something brief about testing the constructors.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/w3c/web-platform-tests/2299)*

**Comment by RByers** [comment]
Reviewed 1 of 3 files at r2, 2 of 2 files at r3.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/w3c/web-platform-tests/2299)*

**Comment by Ms2ger** [comment]
Review status: all files reviewed at latest revision, 9 unresolved discussions.

---

*[touch-events/create-touch-touchlist.html, line 36 \[r3\]](https://reviewable.io:443/reviews/w3c/web-platform-tests/2299#-K6EZj9NDSXFp8MOzlLu:-K6EZj9NDSXFp8MOzlLv:163065205) ([raw file](https://github.com/w3c/web-platform-tests/blob/cfc9ade7ff7043ea24255bdc2908c3f7777542a4/touch-events/create-touch-touchlist.html#L36)):*
Why did you add this? If `createTouch` doesn't exist, the following line will throw and cause a test failure.

---

*[touch-events/touch-touchevent-constructor.html, line 19 \[r3\]](https://reviewable.io:443/reviews/w3c/web-platform-tests/2299#-K6EZUKjnb_nZRZMbhyg:-K6EZUKktNR7sIBW8_dq:940059113) ([raw file](https://github.com/w3c/web-platform-tests/blob/cfc9ade7ff7043ea24255bdc2908c3f7777542a4/touch-events/touch-touchevent-constructor.html#L19)):*
No need for this.

---

*[touch-events/touch-touchevent-constructor.html, line 28 \[r3\]](https://reviewable.io:443/reviews/w3c/web-platform-tests/2299#-K6EZ4KzZiVDXXyvfZ_F:-K6EZ4KzZiVDXXyvfZ_G:-221683987) ([raw file](https://github.com/w3c/web-platform-tests/blob/cfc9ade7ff7043ea24255bdc2908c3f7777542a4/touch-events/touch-touchevent-constructor.html#L28)):*
There doesn't seem to be a reason to wait for the `load` event. Also remove the `setup`/`done` calls.

---

*[touch-events/touch-touchevent-constructor.html, line 30 \[r3\]](https://reviewable.io:443/reviews/w3c/web-platform-tests/2299#-K6EYlK6Wqz6JVSxZs2U:-K6EYlK7ZksCOXjpXrhv:1467327424) ([raw file](https://github.com/w3c/web-platform-tests/blob/cfc9ade7ff7043ea24255bdc2908c3f7777542a4/touch-events/touch-touchevent-constructor.html#L30)):*
Please move those `var`s to where they're used; they're not shared between tests (which would be bad practice).

---

*[touch-events/touch-touchevent-constructor.html, line 65 \[r3\]](https://reviewable.io:443/reviews/w3c/web-platform-tests/2299#-K6EYyqM2tET7Us5giWc:-K6EYyqNAT4kanYF8THC:-911550786) ([raw file](https://github.com/w3c/web-platform-tests/blob/cfc9ade7ff7043ea24255bdc2908c3f7777542a4/touch-events/touch-touchevent-constructor.html#L65)):*
No need for the assignments.

---

*[touch-events/touch-touchevent-constructor.html, line 128 \[r3\]](https://reviewable.io:443/reviews/w3c/web-platform-tests/2299#-K6EZH9PJhDppFvnBEvr:-K6EZH9PJhDppFvnBEvs:306597936) ([raw file](https://github.com/w3c/web-platform-tests/blob/cfc9ade7ff7043ea24255bdc2908c3f7777542a4/touch-events/touch-touchevent-constructor.html#L128)):*
No need to do this.

---

*[touch-events/touch-touchevent-constructor.html, line 142 \[r3\]](https://reviewable.io:443/reviews/w3c/web-platform-tests/2299#-K6EZRIB0QfyA7M4rIBL:-K6EZRIB0QfyA7M4rIBM:-1236376120) ([raw file](https://github.com/w3c/web-platform-tests/blob/cfc9ade7ff7043ea24255bdc2908c3f7777542a4/touch-events/touch-touchevent-constructor.html#L142)):*
Remove the styling

---

*[touch-events/touch-touchevent-constructor.html, line 147 \[r3\]](https://reviewable.io:443/reviews/w3c/web-platform-tests/2299#-K6EZMW_fbGVCO_csbGk:-K6EZMW_fbGVCO_csbGl:-1396021968) ([raw file](https://github.com/w3c/web-platform-tests/blob/cfc9ade7ff7043ea24255bdc2908c3f7777542a4/touch-events/touch-touchevent-constructor.html#L147)):*
Remove the text; it serves no purpose.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/w3c/web-platform-tests/2299)*

**Comment by RByers** [comment]
Review status: all files reviewed at latest revision, 9 unresolved discussions.

---

*[touch-events/touch-touchevent-constructor.html, line 19 \[r3\]](https://reviewable.io:443/reviews/w3c/web-platform-tests/2299#-K6EZUKjnb_nZRZMbhyg:-K6EbZaC1U4L6Cxo-xD9:1364322533) ([raw file](https://github.com/w3c/web-platform-tests/blob/cfc9ade7ff7043ea24255bdc2908c3f7777542a4/touch-events/touch-touchevent-constructor.html#L19)):*
Why do you say that?  We want the tests to be easily usable on mobile and without this you need to do a lot of pinch-zooming to read the test results.  Also note that this is already present in all the other touch event tests.

---

*[touch-events/touch-touchevent-constructor.html, line 28 \[r3\]](https://reviewable.io:443/reviews/w3c/web-platform-tests/2299#-K6EZ4KzZiVDXXyvfZ_F:-K6Ed9UZK1ZapJBYEFLA:-1030551307) ([raw file](https://github.com/w3c/web-platform-tests/blob/cfc9ade7ff7043ea24255bdc2908c3f7777542a4/touch-events/touch-touchevent-constructor.html#L28)):*
Note that Chong was just copying the pattern used by the existing touch tests.  I agree that it's worth cleaning up the test when adding a new one, but presumably at some point we should have someone go back and apply all this feedback to the existing tests this one is based on too.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/w3c/web-platform-tests/2299)*

**Comment by RByers** [comment]
Ms2ger@ thanks for doing a review, we really appreciate it!  Big picture I'd like to get my blink input team (and ultimately the rest of Google web platform teams) contributing much more actively to WPT, so I appreciate your patience while we get ramped up on the process / style.

**Comment by Ms2ger** [comment]
Review status: all files reviewed at latest revision, 9 unresolved discussions.

---

*[touch-events/touch-touchevent-constructor.html, line 19 \[r3\]](https://reviewable.io:443/reviews/w3c/web-platform-tests/2299#-K6EZUKjnb_nZRZMbhyg:-K6EkPOEcLP9ewQYoXFU:1121934435) ([raw file](https://github.com/w3c/web-platform-tests/blob/cfc9ade7ff7043ea24255bdc2908c3f7777542a4/touch-events/touch-touchevent-constructor.html#L19)):*
Looking on my phone, it doesn't seem to improve much? Anyway, we could discuss that outside this PR if you feel strongly.

---

*[touch-events/touch-touchevent-constructor.html, line 28 \[r3\]](https://reviewable.io:443/reviews/w3c/web-platform-tests/2299#-K6EZ4KzZiVDXXyvfZ_F:-K6EklRXDzJhN8Gu9qzL:779216502) ([raw file](https://github.com/w3c/web-platform-tests/blob/cfc9ade7ff7043ea24255bdc2908c3f7777542a4/touch-events/touch-touchevent-constructor.html#L28)):*
I may end up fixing the existing tests at some point, and I don't blame anyone for picking up sub-optimal habits, but I'd like to avoid them in new tests.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/w3c/web-platform-tests/2299)*

**Comment by choniong** [comment]
Reviewed 2 of 3 files at r2.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[touch-events/create-touch-touchlist.html, line 36 \[r3\]](https://reviewable.io:443/reviews/w3c/web-platform-tests/2299#-K6EZj9NDSXFp8MOzlLu:-K6Ekwgafx497FFDKboh:-1137083811) ([raw file](https://github.com/w3c/web-platform-tests/blob/cfc9ade7ff7043ea24255bdc2908c3f7777542a4/touch-events/create-touch-touchlist.html#L36)):*
Because I assume 
`"assert_true: document.createTouch exists expected true got false"`
has more readability than 
`"document.createTouch is not a function
TypeError: document.createTouch is not a function"`
But sure I can remove this.

---

*[touch-events/touch-touchevent-constructor.html, line 28 \[r3\]](https://reviewable.io:443/reviews/w3c/web-platform-tests/2299#-K6EZ4KzZiVDXXyvfZ_F:-K6En0Dc4-pobpxsMST8:-1775589127) ([raw file](https://github.com/w3c/web-platform-tests/blob/cfc9ade7ff7043ea24255bdc2908c3f7777542a4/touch-events/touch-touchevent-constructor.html#L28)):*
Sure I will try to fit them into the right style. Thanks for doing the review!

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/w3c/web-platform-tests/2299)*

**Comment by RByers** [comment]
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[touch-events/touch-touchevent-constructor.html, line 19 \[r3\]](https://reviewable.io:443/reviews/w3c/web-platform-tests/2299#-K6EZUKjnb_nZRZMbhyg:-K6Er0gLZY7ntd5i15V2:391046974) ([raw file](https://github.com/w3c/web-platform-tests/blob/cfc9ade7ff7043ea24255bdc2908c3f7777542a4/touch-events/touch-touchevent-constructor.html#L19)):*
I don't feel particularly strongly on this detail, just that in general WPT tests (especially mobile-focused ones) should be designed to work well on mobile.  How about we keep all the touch tests consistent in this regard for now and follow up in w3c/touch-events#56?

---

*[touch-events/touch-touchevent-constructor.html, line 28 \[r3\]](https://reviewable.io:443/reviews/w3c/web-platform-tests/2299#-K6EZ4KzZiVDXXyvfZ_F:-K6EsHv6lJQHsWujDTl0:-1709418539) ([raw file](https://github.com/w3c/web-platform-tests/blob/cfc9ade7ff7043ea24255bdc2908c3f7777542a4/touch-events/touch-touchevent-constructor.html#L28)):*
@ms2ger agreed, except you shouldn't have to update the tests yourself.  We (TECG members) are happy to do that as long as you (or someone else) can provide guidance / review.  Filed w3c/touch-events##57 to track.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/w3c/web-platform-tests/2299)*

**Comment by RByers** [comment]
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[touch-events/touch-touchevent-constructor.html, line 28 \[r3\]](https://reviewable.io:443/reviews/w3c/web-platform-tests/2299#-K6EZ4KzZiVDXXyvfZ_F:-K6EsrQlyWT2l-TsJ138:-1995276619) ([raw file](https://github.com/w3c/web-platform-tests/blob/cfc9ade7ff7043ea24255bdc2908c3f7777542a4/touch-events/touch-touchevent-constructor.html#L28)):*
Sorry, that's w3c/touch-events#57

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/w3c/web-platform-tests/2299)*

**Comment by choniong** [comment]
Review status: 0 of 3 files reviewed at latest revision, 9 unresolved discussions.

---

*[touch-events/create-touch-touchlist.html, line 36 \[r3\]](https://reviewable.io:443/reviews/w3c/web-platform-tests/2299#-K6EZj9NDSXFp8MOzlLu:-KApyFdjbQMLCQU0DoYT:-499146009) ([raw file](https://github.com/w3c/web-platform-tests/blob/f825b11bb0026a93400a279a763a314f70a8bd3c/touch-events/create-touch-touchlist.html#L36)):*
Done.

---

*[touch-events/touch-touchevent-constructor.html, line 15 \[r3\]](https://reviewable.io:443/reviews/w3c/web-platform-tests/2299#-K6E46sdM8kDaMFmRH7D:-KApyH6ECVLuzjNCBex4:-499146009) ([raw file](https://github.com/w3c/web-platform-tests/blob/f825b11bb0026a93400a279a763a314f70a8bd3c/touch-events/touch-touchevent-constructor.html#L15)):*
Done.

---

*[touch-events/touch-touchevent-constructor.html, line 19 \[r3\]](https://reviewable.io:443/reviews/w3c/web-platform-tests/2299#-K6EZUKjnb_nZRZMbhyg:-KApy7Z1LG62i8-wXGox:-499146009) ([raw file](https://github.com/w3c/web-platform-tests/blob/f825b11bb0026a93400a279a763a314f70a8bd3c/touch-events/touch-touchevent-constructor.html#L19)):*
Done.

---

*[touch-events/touch-touchevent-constructor.html, line 28 \[r3\]](https://reviewable.io:443/reviews/w3c/web-platform-tests/2299#-K6EZ4KzZiVDXXyvfZ_F:-KApy8zkAK3MJZfOSC_F:-499146009) ([raw file](https://github.com/w3c/web-platform-tests/blob/f825b11bb0026a93400a279a763a314f70a8bd3c/touch-events/touch-touchevent-constructor.html#L28)):*
Done.

---

*[touch-events/touch-touchevent-constructor.html, line 30 \[r3\]](https://reviewable.io:443/reviews/w3c/web-platform-tests/2299#-K6EYlK6Wqz6JVSxZs2U:-KApyJ9aLPLaGDO6Kv7k:-499146009) ([raw file](https://github.com/w3c/web-platform-tests/blob/f825b11bb0026a93400a279a763a314f70a8bd3c/touch-events/touch-touchevent-constructor.html#L30)):*
Done.

---

*[touch-events/touch-touchevent-constructor.html, line 65 \[r3\]](https://reviewable.io:443/reviews/w3c/web-platform-tests/2299#-K6EYyqM2tET7Us5giWc:-KApyAEInYNDP7NP_Fd3:-499146009) ([raw file](https://github.com/w3c/web-platform-tests/blob/f825b11bb0026a93400a279a763a314f70a8bd3c/touch-events/touch-touchevent-constructor.html#L65)):*
Done.

---

*[touch-events/touch-touchevent-constructor.html, line 128 \[r3\]](https://reviewable.io:443/reviews/w3c/web-platform-tests/2299#-K6EZH9PJhDppFvnBEvr:-KApyB0TzEUyFP1VkiYe:-499146009) ([raw file](https://github.com/w3c/web-platform-tests/blob/f825b11bb0026a93400a279a763a314f70a8bd3c/touch-events/touch-touchevent-constructor.html#L128)):*
Done.

---

*[touch-events/touch-touchevent-constructor.html, line 142 \[r3\]](https://reviewable.io:443/reviews/w3c/web-platform-tests/2299#-K6EZRIB0QfyA7M4rIBL:-KApyBk50fIomIaSbDEi:-499146009) ([raw file](https://github.com/w3c/web-platform-tests/blob/f825b11bb0026a93400a279a763a314f70a8bd3c/touch-events/touch-touchevent-constructor.html#L142)):*
Done.

---

*[touch-events/touch-touchevent-constructor.html, line 147 \[r3\]](https://reviewable.io:443/reviews/w3c/web-platform-tests/2299#-K6EZMW_fbGVCO_csbGk:-KApyC4QVyG-AtoU4HfY:-499146009) ([raw file](https://github.com/w3c/web-platform-tests/blob/f825b11bb0026a93400a279a763a314f70a8bd3c/touch-events/touch-touchevent-constructor.html#L147)):*
Done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/w3c/web-platform-tests/2299)*

**Comment by choniong** [comment]
Hi @Ms2ger , sorry for the slow response, but can you take a look at the updated PR please? Thanks!

File: touch-events/touch-touchevent-constructor.html [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+<title>Touch and TouchEvent Constructor Tests</title>
+<script src="/resources/testharness.js"></script>
+<script src="/resources/testharnessreport.js"></script>
+<script src="touch-support.js"></script>
+</head>
+<body>
+<div id="target0"></div>
+<script>
+test(function() {
+    assert_throws({name: 'TypeError'}, function() {new Touch();}, "Touch constructor requires initialize dictionary");
```
Comment (Ms2ger) [inline touch-events/touch-touchevent-constructor.html]: `assert_throws(new TypeError(), ...)` is more idiomatic

**Comment by Ms2ger** [comment]
On the testharness side, this looks good to me with the one nit fixed. No comment on whether the tests are correct otherwise.

**Comment by choniong** [comment]
Thanks for the prompt response! I've fixed the nit, so is it ok to merge or do I need to ask for another correctness review?

**Comment by RByers** [comment]
> Thanks for the prompt response! I've fixed the nit, so is it ok to merge or do I need to ask for another correctness review?

I looked over the correctness again, and it still looks good to me.  Please run your test on Chrome after your final changes (if you haven't already) to verify it still passes.  Then I'm guessing @Ms2ger will want you to squash all your commits into a single one (use `git rebase -i`) before he merges.

**Comment by choniong** [comment]
Sure I've tested on Chrome and it still passes, and commits are squashed into one now. Thanks!

**Comment by jdm** [comment]
Thank you @choniong!

**Pull request merged**